sw/device/lib/dif/dif_kmac_unittest.cc

```diff
@@ -86,5 +89,94 @@ TEST(FunctionNameTest, BadArg) {
   EXPECT_EQ(dif_kmac_function_name_init("", -1, &fn), kDifBadArg);
 }
 
+using mock_mmio::MmioTest;
+using mock_mmio::MockDevice;
+using testing::Test;
+
+// Base class for the rest fixtures in this file.
+class KmacTest : public testing::Test, public mock_mmio::MmioTest {
+ public:
+  dif_kmac_t kmac_;
+  dif_kmac_operation_state_t kmac_operation_state_;
+
+  KmacTest() { EXPECT_EQ(dif_kmac_init(dev().region(), &kmac_), kDifOk); }
+
+  /**
+   * Set mmio write expectation for 8 bits data size.
+   *
+   * @param message Buffer with the data.
+   * @param size Len of the buffer.
+   */
+  void setExpectedMessageByte(const uint8_t *message, const size_t size = 8) {
+    for (int i = 0; i < size; ++i) {
+      EXPECT_WRITE8(KMAC_MSG_FIFO_REG_OFFSET, message[i]);
+    }
+  }
+
+  /**
+   * Set mmio write expectation for 32 bits data size considering an alignment
+   * of 32 bits.
+   *
+   * @param message Buffer with the data.
+   * @param size Len of the buffer.
+   */
+  void setExpectedMessageInt32(const uint8_t *message, const size_t size = 8) {
+    // Check if the buffer is unaligned
+    size_t remaining = size;
+    size_t unalignment = ((uintptr_t)message) % sizeof(uint32_t);
+    if (unalignment) {
+      // write unaligned data in bytes
+      setExpectedMessageByte(message, unalignment);
+      message += unalignment;
+      remaining -= unalignment;
+    }
+
+    // Write aligned part of the buffer
+    while (remaining > 3) {
+      uint32_t word = 0;
+      memcpy(&word, message, sizeof(uint32_t));
+      EXPECT_WRITE32(KMAC_MSG_FIFO_REG_OFFSET, word);
+      remaining -= sizeof(uint32_t);
+      message += sizeof(uint32_t);
+    }
+    // Check if there still unaligned data in the buffer tail
+    if (remaining > 0) {
+      // write unaligned data in bytes
+      setExpectedMessageByte(message, remaining);
+    }
+  }
+};
+
+TEST_F(KmacTest, AbsorbAlignedMessage) {
+  __attribute__((aligned(sizeof(uint32_t)))) static const uint8_t kMessage[] = {
```

Comment:
I personally probably would move this and the unaligned buffer out of this function and put it together with the class variables, but it's up to you.